### ISSUE TITLE
Update Event usage and remove DOMError in favour of e.g. MediaRecorderErrorEvent

### ISSUE THREAD

**Issue by yell0wd0g** [w3c/mediacapture-record#31]
During https://github.com/w3c/mediacapture-record/pull/29, foolip@ had some comments about the Events and DOMError. (Seems like a similar discussion took place in https://github.com/w3c/mediacapture-record/issues/2 but did not get far.

foolip quote (https://github.com/w3c/mediacapture-record/pull/29#issuecomment-162528440): 

> Had a look at the changes and commented in a few places. High-level, I think the biggest issues around events and exceptions are these:
> - Events are fired synchronously inside various algorithms, like the pause event in the pause() method. This is probably not intended, and not how it's implemented in Blink. Should probably say "queue a task to ..." and define what the task source for those tasks is, and the "DOM manipulation task source" would probably be fine. (Probably some of the methods need to return and run in parallel.)
> - DOMError has been removed from DOM, and it was not an Event interface. For async errors, a simple event with name "error" plus an error attribute, or a custom error event interface like MediaRecorderErrorEvent, seems necessary.
> - The verb "raise" is used for both exceptions and events. The right terminology is e.g. "throw an InvalidStateError exception" and "fire an event named dataavailable".

And FTR, Gecko already uses a [RecordErrorEvent](https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/RecordErrorEvent.webidl), whereas Blink needs update

**Comment by yell0wd0g** [comment]
foolip@ also commented in https://github.com/w3c/mediacapture-record/issues/14:

> I would probably suggest firing a simple error event, and adding a readonly attribute DOMException error to the MediaRecorder interface instead. This is how HTMLMediaElement and IDBRequest etc. do similar things.

**Comment by yell0wd0g** [comment]
FTR and to keep information in this bug/issue. Gecko defines a RecordErrorEvent [1] that is used in MediaRecorder.cpp [2] by filling its `name` in `NotifyError()` to e.g. `SecurityError`, `OutOfMemoryError` or otherwise `GenericError`. To my naked eye there is no hard link between these strings and the entries in  `RecordingErrorEventEnum`.

[1] https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/RecordErrorEvent.webidl
[2] https://github.com/mozilla/gecko-dev/blob/master/dom/media/MediaRecorder.cpp

**Comment by yell0wd0g** [comment]
Finally, IMHO we could also try a CustomEvent (https://dom.spec.whatwg.org/#interface-customevent) and stuff the error type/string into its `detail` field.

**Comment by foolip** [comment]
In https://github.com/w3c/mediacapture-record/issues/2#issuecomment-57291319 @alvestrand said that "This document should be following the same pattern as Media Capture and Streams for its errors."

The [Media Capture and Streams spec](http://w3c.github.io/mediacapture-main/) introduces a OverconstrainedError and an ErrorEvent interface.

The closest thing to that would be keep MediaRecorderErrorEvent, I think.

The most important concern here, I think, is what changes are deemed to be safe for Gecko, which has already shipped this API. Who is their implementor?

CC @annevk and @domenic, who in general have much to say about ES6 errors, DOMError, DOMException, etc.

**Comment by alvestrand** [comment]
The OverconstrainedError is the special case - it was introduced (with much pain and copying from the Javascript spec) because we had to carry more information than the name of the error.
In all other cases, Media Capture and Streams went with using an ErrorEvent, which contains an Error (which in turn contains an error name and a free-form "message" field).
Unless you really, really have to carry more information than the error name, I recommend using the ErrorEvent only.

**Comment by yell0wd0g** [comment]
Just FTR, [MediaStreamError](http://www.w3.org/TR/mediacapture-streams/#mediastreamerror) 

``` [NoInterfaceObject]
interface MediaStreamError {
    readonly    attribute DOMString  name;
    readonly    attribute DOMString? message;
    readonly    attribute DOMString? constraintName;
};
```

is very similar to [Gecko](https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/RecordErrorEvent.webidl#L7)

```
interface RecordErrorEvent : Event
{
  readonly attribute DOMString name;
}
```

I'll propose a patch with a `name` and an `message`.

**Comment by mreavy** [comment]
I like this; it's a smart plan to synchronize the specs.  Thanks

**Comment by alvestrand** [comment]
Note: The current media capture and streams error handling uses a naked Error.

http://w3c.github.io/mediacapture-main/#error-handling

**Comment by beverloo** [comment]
As a random question - are you aware that there are two ErrorEvent interfaces?

https://html.spec.whatwg.org/multipage/webappapis.html#the-errorevent-interface
http://w3c.github.io/mediacapture-main/#errorevent (duplicated [here](https://rawgit.com/w3c/mediacapture-record/master/MediaRecorder.html#errorevent)— could we introduce a dependency instead?)

This seems unfortunate. The media one is a subset of the HTML one, compatible due to the nature of `any error` in both the interface and the init dictionary.

On a general note, I don't think throwing `error` events with no information or whatsoever about _what went wrong_ is a great position to place a Web developer in, especially if, in this case, such errors are fatal and cause recording to stop. Perhaps something to fix across the media specs?

Section 7 of the Media Recorder specification still notes the following:

> A fatal error has occurred and the UA has stopped recording. More detailed error information is available in the 'message' attribute.

This is not true anymore.

**Comment by yell0wd0g** [comment]
@beverloo, I agree that is meh -at best- to just send a naked error with no extra information. I wanted to implement the changes in two steps: firstly remove the MediaRecorderErrorEvent in favour of an aligned simple `error`, then  discuss what's the best way to move forward. Note that Cr/Blink does not produce any meaningful Errors (yet).

I believe @foolip already enumerated the options:
- add an `error` attribute, see https://github.com/w3c/mediacapture-record/issues/31#issuecomment-162703378:  HTMLMediaElement implements a naked `Error Event` with an [`error` attribute](https://html.spec.whatwg.org/#dom-media-error)
- use an (ErrorEvent)[https://html.spec.whatwg.org/multipage/webappapis.html#the-errorevent-interface]  which has the necessary `message` field to add more information as to what's the condition, and is widely used in Blink (but I'm not super familiar with this API so I might be perusing it the wrong way)
- any others...?

I would like to hear the feedback from Gecko which has the most mature implementation of this, @mreavy?

**Comment by mreavy** [comment]
I believe we should work to unify error handling, at least for new APIs, across media, and we should use ErrorEvents with messages.

**Comment by foolip** [comment]
Note that the "naked Error" in http://w3c.github.io/mediacapture-main/#error-handling and the "naked Error Event" for media elements are not the same thing.

The media capture spec does some kind of Error subclassing, while media elements (HTML) says "fire a simple event named error", which means using the `Event` interface (not `Error` or `EventError`). The actual details about what went wrong with media elements is in the `MediaError` instance then found on `mediaElement.error`.

This is all problematic in HTML as well, and there are a few open issues:
https://github.com/whatwg/html/issues/346
https://github.com/whatwg/html/issues/505

**Comment by jan-ivar** [comment]
@foolip note that [`OverconstrainedError`](http://w3c.github.io/mediacapture-main/#dfn-overconstrainederror), the lone error subclass in media capture, has a dedicated [`OverconstrainedErrorEvent`](http://w3c.github.io/mediacapture-main/#idl-def-OverconstrainedErrorEvent), so unless I'm missing something, I believe we're fine to use the html `EventError`, and I've filed https://github.com/w3c/mediacapture-main/issues/335.

**Comment by foolip** [comment]
Thanks, I've commented on the new issue.

**Comment by yell0wd0g** [comment]
Closing to continue the conversation in #53
